File: frontend/src/types/rubricAlignment.ts
/**
 * STEP 28: Answer <-> Rubric Alignment types.
 *
 * Alignment describes how well an answer addresses each rubric criterion. It is not a grade
 * and not correctness; it never changes faculty marks.
 */

export type AlignmentStatus = 'STRONG' | 'PARTIAL' | 'WEAK' | 'NOT_ALIGNED';

export type AlignmentAnalysisStatus = 'PENDING' | 'PROCESSING' | 'COMPLETED' | 'FAILED' | 'REVIEWED' | 'STALE';

export const ALIGNMENT_WEIGHTS: Record<AlignmentStatus, number> = {
  STRONG: 1,
  PARTIAL: 0.5,
  WEAK: 0.25,
  NOT_ALIGNED: 0,
};

export interface AlignmentCounts {
  strong: number;
  partial: number;
  weak: number;
  not_aligned: number;
}

export interface CriterionAlignment {
  id: number;
  rubric_criterion_id: number | null;
  criterion: string;
  max_marks: number;
  alignment_score: number;
  similarity?: number | null;
  alignment_status: AlignmentStatus;
  evidence: string[];
  missing_elements: string[];
  explanation: string;
  sort_order?: number;
}

export interface RubricAlignment {
  id: number;
  student_answer_id: number;
  student_submission_id: number;
  question_id: number;
  rubric_id: number | null;
  rubric_version: number | null;
  /** Mark-weighted alignment percentage (primary). */
  overall_alignment_score: number | null;
  unweighted_alignment_score: number | null;
  alignment_status: AlignmentStatus | null;
  analysis_status: AlignmentAnalysisStatus;
  is_current: boolean;
  is_stale: boolean;
  stale_reasons: string[];
  counts: AlignmentCounts;
  summary: string | null;
  strengths: string[];
  missing_elements: string[];
  error_message: string | null;
  model_name?: string | null;
  model_version?: string | null;
  analysis_method?: string | null;
  thresholds?: Record<string, number> | null;
  generated_at: string | null;
  reviewed_at?: string | null;
  reviewed_by?: number | null;
  created_at?: string;
  updated_at?: string;
  criterion_alignments: CriterionAlignment[];
}

export const isAlignmentActive = (status: AlignmentAnalysisStatus | null | undefined): boolean =>
  status === 'PENDING' || status === 'PROCESSING';

export const isAlignmentCompleted = (status: AlignmentAnalysisStatus | null | undefined): boolean =>
  status === 'COMPLETED' || status === 'REVIEWED' || status === 'STALE';

/** Gap (percentage points) between AI suggested marks and alignment beyond which a review notice is shown. */
export const ALIGNMENT_GRADE_GAP_THRESHOLD = 30;

/**
 * Compare the STEP 27 suggested-marks percentage with the STEP 28 alignment percentage.
 * Returns the gap in points, or null when either signal is unavailable. Never used to change a grade.
 */
export function gradingAlignmentGap(
  suggestedMarks: number | null | undefined,
  maximumMarks: number | null | undefined,
  alignmentScore: number | null | undefined,
): number | null {
  if (suggestedMarks === null || suggestedMarks === undefined) return null;
  if (!maximumMarks || maximumMarks <= 0) return null;
  if (alignmentScore === null || alignmentScore === undefined) return null;
  return Math.round(((suggestedMarks / maximumMarks) * 100 - alignmentScore) * 10) / 10;
}

File: ai-service/app/schemas/rubric_alignment.py
"""Pydantic schemas for STEP 28 Answer <-> Rubric Alignment.

Alignment describes how well an answer *addresses* each rubric criterion. It is
not correctness and it is not a grade. Faculty review remains necessary.
"""

from typing import List, Optional

from pydantic import BaseModel, Field, model_validator

from app.schemas.grading import GradingAnswerIn, GradingCriterionIn
from app.schemas.rubric import CourseContext, LearningOutcomeContext

ALIGNMENT_STATUSES = ("STRONG", "PARTIAL", "WEAK", "NOT_ALIGNED")
ALIGNMENT_WEIGHTS = {"STRONG": 1.0, "PARTIAL": 0.5, "WEAK": 0.25, "NOT_ALIGNED": 0.0}
MARK_TOLERANCE = 0.005
SCORE_TOLERANCE = 0.05


class AlignmentQuestionIn(BaseModel):
    id: Optional[int] = None
    text: str = Field(..., min_length=1, max_length=10000)
    total_marks: Optional[float] = Field(default=None, gt=0, le=1000)
    question_type: Optional[str] = Field(default=None, max_length=30)
    expected_answer: Optional[str] = Field(default=None, max_length=10000)


class AlignmentRubricIn(BaseModel):
    id: Optional[int] = None
    version: Optional[int] = None
    total_marks: Optional[float] = Field(default=None, gt=0)
    criteria: List[GradingCriterionIn] = Field(..., min_length=1, max_length=12)

    @model_validator(mode="after")
    def criteria_must_be_consistent(self):
        ids = [c.id for c in self.criteria]
        if len(ids) != len(set(ids)):
            raise ValueError("Rubric criteria identifiers must be unique.")
        total = round(sum(c.max_marks for c in self.criteria), 2)
        if total <= 0:
            raise ValueError("Rubric criteria must carry marks greater than zero.")
        if self.total_marks is not None and abs(total - round(self.total_marks, 2)) > MARK_TOLERANCE:
            raise ValueError(
                f"Rubric criterion marks ({total}) do not sum to the rubric total ({self.total_marks})."
            )
        return self


class AnalyzeAnswerRubricAlignmentRequest(BaseModel):
    student_answer: GradingAnswerIn
    question: AlignmentQuestionIn
    rubric: AlignmentRubricIn
    learning_outcome: Optional[LearningOutcomeContext] = None
    course_context: Optional[CourseContext] = None


class CriterionAlignmentOut(BaseModel):
    rubric_criterion_id: int
    criterion: str = Field(..., min_length=1, max_length=255)
    max_marks: float = Field(..., ge=0)
    alignment_status: str
    alignment_score: float = Field(..., ge=0, le=1, description="Deterministic weight for the status (1 / .5 / .25 / 0)")
    similarity: float = Field(..., ge=0, le=1, description="Combined semantic/lexical criterion signal used for classification")
    evidence: List[str] = Field(default_factory=list)
    missing_elements: List[str] = Field(default_factory=list)
    explanation: str = Field(..., min_length=1, max_length=2000)

    @model_validator(mode="after")
    def status_and_weight_consistent(self):
        if self.alignment_status not in ALIGNMENT_STATUSES:
            raise ValueError(f"Unknown alignment status '{self.alignment_status}'.")
        if abs(ALIGNMENT_WEIGHTS[self.alignment_status] - self.alignment_score) > 1e-6:
            raise ValueError("alignment_score does not match the alignment_status weight.")
        return self


class AlignmentCounts(BaseModel):
    strong: int = 0
    partial: int = 0
    weak: int = 0
    not_aligned: int = 0


class AlignmentMetadata(BaseModel):
    model: str
    version: str = "1.0.0"
    embedding_model: Optional[str] = None
    generative_model_used: bool = False
    method: str = "semantic_and_rubric_alignment"
    thresholds: dict = Field(default_factory=dict)
    criteria_count: int = 0
    validation_passed: bool = True
    disclaimer: str = (
        "AI-generated rubric alignment is an assistive analysis. It may miss context, nuance, "
        "or valid alternative answers. Faculty review remains necessary."
    )


class AnalyzeAnswerRubricAlignmentResponse(BaseModel):
    status: str = "success"
    overall_alignment_score: float = Field(..., ge=0, le=100, description="Mark-weighted alignment percentage (primary)")
    unweighted_alignment_score: float = Field(..., ge=0, le=100)
    overall_alignment_status: str
    counts: AlignmentCounts
    summary: str = Field(..., min_length=1, max_length=2000)
    strengths: List[str] = Field(default_factory=list)
    missing_elements: List[str] = Field(default_factory=list)
    criterion_alignments: List[CriterionAlignmentOut] = Field(..., min_length=1)
    metadata: AlignmentMetadata

    @model_validator(mode="after")
    def scores_must_be_consistent(self):
        if self.overall_alignment_status not in ALIGNMENT_STATUSES:
            raise ValueError(f"Unknown overall alignment status '{self.overall_alignment_status}'.")
        total_marks = sum(c.max_marks for c in self.criterion_alignments)
        if total_marks > 0:
            weighted = round(sum(c.alignment_score * c.max_marks for c in self.criterion_alignments) / total_marks * 100, 2)
            if abs(weighted - self.overall_alignment_score) > SCORE_TOLERANCE:
                raise ValueError(
                    f"overall_alignment_score ({self.overall_alignment_score}) does not match the mark-weighted criterion scores ({weighted})."
                )
        unweighted = round(sum(c.alignment_score for c in self.criterion_alignments) / len(self.criterion_alignments) * 100, 2)
        if abs(unweighted - self.unweighted_alignment_score) > SCORE_TOLERANCE:
            raise ValueError("unweighted_alignment_score does not match the criterion scores.")
        return self
